internal/locate/replica_selector.go

```diff
@@ -0,0 +1,554 @@
+		isStaleRead:     req.StaleRead,
+		isReadOnlyReq:   isReadReq(req.Type),
+		option:          option,
+		target:          nil,
+		attempts:        0,
+	}, nil
+}
+
+func (s *replicaSelectorV2) next(bo *retry.Backoffer, req *tikvrpc.Request) (rpcCtx *RPCContext, err error) {
+	if !s.region.isValid() {
+		metrics.TiKVReplicaSelectorFailureCounter.WithLabelValues("invalid").Inc()
+		return nil, nil
+	}
+
+	s.attempts++
+	s.target = nil
+	s.proxy = nil
+	switch s.replicaReadType {
+	case kv.ReplicaReadLeader:
+		s.nextForReplicaReadLeader(req)
+	default:
+		s.nextForReplicaReadMixed(req)
+	}
+	if s.target == nil {
+		return nil, nil
+	}
+	return s.buildRPCContext(bo, s.target, s.proxy)
+}
+
+func (s *replicaSelectorV2) nextForReplicaReadLeader(req *tikvrpc.Request) {
+	if s.regionCache.enableForwarding {
+		strategy := ReplicaSelectLeaderWithProxyStrategy{}
+		s.target, s.proxy = strategy.next(s.replicas, s.region)
+		if s.target != nil && s.proxy != nil {
+			return
+		}
+	}
+	leaderIdx := s.region.getStore().workTiKVIdx
+	strategy := ReplicaSelectLeaderStrategy{leaderIdx: leaderIdx}
+	s.target = strategy.next(s.replicas)
+	if s.target != nil && s.busyThreshold > 0 && s.isReadOnlyReq && (s.target.store.EstimatedWaitTime() > s.busyThreshold || s.target.serverIsBusy) {
+		// If the leader is busy in our estimation, try other idle replicas.
+		// If other replicas are all busy, tryIdleReplica will try the leader again without busy threshold.
+		mixedStrategy := ReplicaSelectMixedStrategy{leaderIdx: leaderIdx, busyThreshold: s.busyThreshold}
+		idleTarget := mixedStrategy.next(s, s.region)
+		if idleTarget != nil {
+			s.target = idleTarget
+			req.ReplicaRead = true
+		} else {
+			// No threshold if all peers are too busy, remove busy threshold and still use leader.
+			s.busyThreshold = 0
+			req.BusyThresholdMs = 0
+			req.ReplicaRead = false
+		}
+	}
+	if s.target != nil {
+		return
+	}
+	mixedStrategy := ReplicaSelectMixedStrategy{leaderIdx: leaderIdx, leaderOnly: s.option.leaderOnly}
+	s.target = mixedStrategy.next(s, s.region)
+	if s.target != nil && s.isReadOnlyReq && s.replicas[leaderIdx].deadlineErrUsingConfTimeout {
+		req.ReplicaRead = true
+		req.StaleRead = false
+	}
+}
+
+func (s *replicaSelectorV2) nextForReplicaReadMixed(req *tikvrpc.Request) {
+	leaderIdx := s.region.getStore().workTiKVIdx
+	if s.isStaleRead && s.attempts == 2 {
+		// For stale read second retry, try leader by leader read.
+		strategy := ReplicaSelectLeaderStrategy{leaderIdx: leaderIdx}
+		s.target = strategy.next(s.replicas)
+		if s.target != nil && !s.target.isExhausted(1, 0) {
+			// For stale read, don't retry leader again if it is accessed at the first attempt.
+			req.StaleRead = false
+			req.ReplicaRead = false
+			return
+		}
+	}
+	preferLeader := req.ReplicaReadType == kv.ReplicaReadPreferLeader
+	if s.attempts > 1 {
+		if req.ReplicaReadType == kv.ReplicaReadMixed {
+			// For mixed read retry, prefer retry leader first.
+			preferLeader = true
+		}
+	}
```

Comment:
No need to do this, since normally, in `ReplicaSelectMixedStrategy.isCandidate`, the replica `maxAttempt` is 1, so if the leader is attempted, then it won't be attempted again.